Add instructions to configure HPA on openshift (#149)
3scale/apisonator: docs/openshift_horizontal_scaling.md

# Openshift Horizontal Pod Autoscaling (HPA)

> It is possible to configure Apisonator to auto scale when deployed in Openshift.

Reviewer: nit:
```suggestion
It is possible to configure Apisonator to automatically scale when deployed on Openshift.
```

Reviewer: perhaps mention this can also be done in bare k8s?

Author: Done :+1:

guide](https://github.com/3scale/3scale-operator/blob/master/doc/template-user-guide.md).

## Openshift HPA

Openshift Horizontal Pod Autoscaling (HPA) allows to scale pods horizontally by
specifying the minimum and maximum number of pods that should be deployed and
the target CPU utilization. For more details check the [official Openshift
docs](https://docs.openshift.com/container-platform/4.1/nodes/pods/nodes-pods-autoscaling.html).

## Async

Apisonator can be configured to use a non-blocking Redis client. With this
feature enabled, a single process should use a whole CPU when properly tuned.
This means that we can enable Openshift HPA using the CPU utilization as the
scaling metric.

To learn more about how the async Redis client works, check this [design
doc](./async.md).

## Setup

### Redis databases

The standard 3scale deployment uses a single Redis pod and sets both the resque
DB and the data DB in the same Redis process under different DB indexes. This is
not supported by the async redis library, so we need to either place both DBs in
the same Redis process or use two separate ones. To configure this, change the

> `CONFIG_REDIS_PROXY` and `REDIS_QUEUES_URL` envs in the listener and worker

Author: This variable was introduced a long time ago. We can discuss ENV name changes in a separate PR because they require other changes to ensure backwards compatibility.

deployment configs.

### Enable async

- Set the `CONFIG_REDIS_ASYNC` to true in the listener and worker deployment configs.

> - Configure listeners to use the Falcon web-server instead of Puma. In order to

Reviewer: This depends on #150.

Author: Yes, go approve it :smile:

> do this, modify the command in the listener deployment config. Replace this:

Author: I agree. In the future, this should be implemented in the operator, but I wanted to document all the apisonator-specific knowledge here and validate the idea first.

```yaml
spec:
  containers:
    - args:
        - bin/3scale_backend
        - start
        - '-e'
        - production
        - '-p'
        - '3000'
        - '-x'
        - /dev/stdout
```

with:
```yaml
spec:
  containers:
    - args:
        - bin/3scale_backend
        - '-s'
        - falcon
        - start
        - '-e'
        - production
        - '-p'
        - '3000'
        - '-x'
        - /dev/stdout
```

> - Set the `PUMA_WORKERS` env to 1 in the listener deployment. This will deploy 1

Reviewer: How about introducing `LISTENER_WORKERS`, look it up with higher precedence in the code, and if not present fall back to `PUMA_WORKERS`?

Falcon worker per pod.

### Set up horizontal pod autoscaling

We need to decide the target CPU. It is defined as a percentage of the requested

> CPU. In a default installation, the backend-listener pod requests 0.5 CPUs. In

Author: I added some changes to clarify this.

requests 0.5 CPU cores. In an ideal case, we know that both the listener and the
worker should use 1 CPU core when using the async redis lib. However, we need to
leave some margin for possible inefficiencies, so we can target 0.8 CPU cores,
for example. 0.8/0.5 = 1.6, so we need to define a CPU % of 160 for the
listener. In a default installation, the backend-worker pod requests 0.150 CPU
cores. Following the same logic, we need to define a target CPU of 533
(0.8/0.150 = 5.33). Pay attention to the CPU request of each pod, as they might

> change between versions, and adapt the target CPU % values accordingly.

Reviewer: So here comes the logical question: why would we want to scale so much more the workers CPU demand (to 0.8) when we only request initially 0.150 millicores (when compared to the listeners which start way higher)? We could add something, even if a little vague, about why this is the case.

Author: Even if you request only 0.150, you can ask for a limit of 1.

Author: In my opinion, it's a bit confusing to define the target CPU as a function of the pod request. It'd be more intuitive to just define the number of cores.

Reviewer: And the docs are not helping. Will look them up in k8s docs and see if there is any clearer explanation.

We also need to define the minimum and the maximum number of pods. The example
below uses 1 and 10, but adapt those values according to the resources available
in your Openshift cluster.

Configure HPA both in listeners and workers:
```bash
oc autoscale dc/backend-listener --min 1 --max 10 --cpu-percent=160
oc autoscale dc/backend-worker --min 1 --max 10 --cpu-percent=533
```

To check the decisions made by the auto-scaler as well as information about the
current number of pods vs the desired number:
```bash
oc describe hpa backend-listener
oc describe hpa backend-worker

> ```

Reviewer 2: Add a little bit about seeing it work when testing, load increases?
